net/libfab/libfab.c

```diff
@@ -0,0 +1,2886 @@
+			if (tm->ftm_shutdown)
+				break;
+			
+			if (libfab_tm_trylock(tm) != 0) {
+				m0_mutex_unlock(&tm->ftm_endlock);
+				/* the tm can not be locked now. 
+				   We will try it again */
+			} else {
+				/* Got tm lock. 
+				   Let's continue processing events. */
+				break;
+			}
+		}
+		
+		m0_mutex_unlock(&tm->ftm_endlock);
+		
+		if (tm->ftm_shutdown)
+			break;
+		
+		M0_ASSERT(libfab_tm_is_locked(tm) && libfab_tm_invariant(tm));
+
+		if (ev_cnt > 0) {
+			libfab_handle_connect_request_events(tm);
+			libfab_txep_comp_read(tm->ftm_tx_cq);
+
+			m0_tl_for(m0_nep, &tm->ftm_ntm->ntm_end_points, net) {
+				xep = libfab_ep(net);
+				aep = libfab_aep_get(xep);
+				if (aep != NULL) {
+					libfab_txep_event_check(xep, aep, tm);
+					cq = aep->aep_rx_res.frr_cq;
+					libfab_rxep_comp_read(cq, xep);
+				}
+			} m0_tl_endfor;
+		}
+
+		libfab_bulk_buf_process(tm);
+		libfab_tm_buf_timeout(tm);
+		libfab_tm_buf_done(tm);
+
+		M0_ASSERT(libfab_tm_invariant(tm));
+		libfab_tm_unlock(tm);
+	}
+}
+
+/** 
+ * Converts network end-point to its libfabric structure.
+ */
+static inline struct m0_fab__ep *libfab_ep(struct m0_net_end_point *net)
+{
+	return container_of(net, struct m0_fab__ep, fep_nep);
+}
+
+/**
+ * Compares the endpoint name with the passed name string and
+ * returns true if equal, or else returns false
+ * If name is null then the ipaddress and port fields are compared to check for
+ * a matching endpoint.
+ */
+static bool libfab_ep_cmp(struct m0_fab__ep *ep, const char *name,
+			  struct m0_fab__ep_name *epname)
+{
+	bool ret = false;
+	M0_PRE(ep != NULL);
+	if (name == NULL) {
+		M0_ASSERT(epname != NULL);
+		if ((strcmp(ep->fep_name.fen_addr, epname->fen_addr) == 0) &&
+		    (strcmp(ep->fep_name.fen_port, epname->fen_port) == 0))
+			ret = true;
+	} else {
+		if (strcmp(ep->fep_name.fen_str_addr, name) == 0)
+			ret = true;
+	}
+	return ret;
+}
+
+/**
+ * Search for the ep in the existing ep list using one of the following -
+ *   1) Name in str format      OR
+ *   2) ipaddr and port 
+ * If found then return the ep structure, or else create a new endpoint
+ * with the name
+ * Returns 0 if the endpoint is found/created succesfully,
+ * else returns error code.
+ */
+static int libfab_ep_find(struct m0_net_transfer_mc *tm, const char *name,
+			  struct m0_fab__ep_name *epn,
+			  struct m0_net_end_point **epp)
+{
+	struct m0_net_end_point  *net;
+	struct m0_fab__ep        *ep;
+	struct m0_fab__active_ep *aep;
+	struct m0_fab__tm        *ma;
+	char                      ep_str[LIBFAB_ADDR_STRLEN_MAX] = {'\0'};
+	char                     *wc = NULL;
+	bool                      found = false;
+	int                       rc = 0;
+
+	M0_ASSERT(libfab_tm_is_locked(tm->ntm_xprt_private));
+	m0_tl_for(m0_nep, &tm->ntm_end_points, net) {
+		ep = libfab_ep(net);
+		if (libfab_ep_cmp(ep, name, epn)) {
+			*epp = &ep->fep_nep;
+			found = true;
+			libfab_ep_get(ep);
+			break;
+		}
+	} m0_tl_endfor;
```

Comment:
m0_tl_find() can be used here. most likely refactoring to tl_find() will make the code cleaner.